### ISSUE TITLE
[6.2.1 client, macOS, arm64] cannot use AES encryption: `Error: cannot import python-cryptography`

### ISSUE THREAD

**Issue by psobolewskiPhD** [Xpra-org/xpra#4408]
**Describe the bug**
Following the AES example usage (https://github.com/Xpra-org/xpra/blob/master/docs/Network/AES.md) trying to attach with 6.2.1 client on macOS arm64 (DMG downloaded from https://github.com/Xpra-org/xpra/tree/master#official-stable-downloads ) the connection fails with a traceback:

```
2024-10-31 16:52:05,987 Error: cannot import python-cryptography
2024-10-31 16:52:05,987 error preparing connection: invalid encryption 'AES', options: none
Traceback (most recent call last):
  File "/Users/sobolp/Applications/Xpra.app/Contents/Resources/lib/python/xpra/client/base/client.py", line 379, in send_hello
    hello = self.make_hello_base()
            ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sobolp/Applications/Xpra.app/Contents/Resources/lib/python/xpra/client/base/client.py", line 455, in make_hello_base
    cipher_caps = self.get_cipher_caps()
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sobolp/Applications/Xpra.app/Contents/Resources/lib/python/xpra/client/base/client.py", line 472, in get_cipher_caps
    raise ValueError(f"invalid encryption {enc!r}, options: {csv(ciphers) or 'none'}")
ValueError: invalid encryption 'AES', options: none
```

If I don't use AES, everything works, but the statistics tab also results in spam `Error: cannot import python-cryptography`

**To Reproduce**
1. server command: `xpra start --start=xterm \
     --bind-tcp=0.0.0.0:10000,encryption=AES,keyfile=aes.txt` (xpra v6.2.1-r0, ubuntu jammy, from official source)
2. client command: `xpra attach "tcp://<ip>:10000/?encryption=AES&keyfile=./aes.txt"`
3. specific action to trigger the bug: None, can't connect

**System Information (please complete the following information):**
 - Server OS: [e.g. RHEL 8.x] Ubuntu Jammy
 - Client OS: [e.g. Windows 10] MacOS 14.7
 - Xpra Server Version [e.g. 4.2.3] : xpra v6.2.1-r0
 - Xpra Client Version [e.g. 4.3] : v6.2.1-r0

**Additional context**
This used to work in March, so using xpra 5.0, although I needed to set: `CRYPTOGRAPHY_OPENSSL_NO_LEGACY=1`

**Comment by totaam** [comment]
Something must have changed in the latest version of `python-cryptography` or one of its dependencies.
Which means we also have to update the packaging scripts - and for all xpra versions, as this will affect newer v5.x builds just the same.

Here are the recent updates:
* [42.0.4](https://github.com/Xpra-org/gtk-osx-build/commit/b14caccd9632ba71a7097b718a5550c1b2ed55c6)
* [42.0.5](https://github.com/Xpra-org/gtk-osx-build/commit/de5e318c128d36eb37ecd59e2ed7b4ba3d8e04ed)
* [42.0.7](https://github.com/Xpra-org/gtk-osx-build/commit/030e055965b1ec864faedec1b14d26810b28a138)
* [42.0.8](https://github.com/Xpra-org/gtk-osx-build/commit/7db523518e1c12fcb02a4412a22f9246d4c4c6ae)
* [43.0.0](https://github.com/Xpra-org/gtk-osx-build/commit/eb989c9a18d2deb8e14c03377937b4436d789a98)
* [43.0.1](https://github.com/Xpra-org/gtk-osx-build/commit/dd39bb9da230245bfdebe6b38086d01c1cb7e827)
* [43.0.3](https://github.com/Xpra-org/gtk-osx-build/commit/81ff5d17e825296e110fbe23b9c709dd1e98e86a)

**Comment by totaam** [comment]
I can reproduce the problem simply by running `Xpra.app/Contents/Helpers/Network_info -v`.
```py
crypto backend init failure
Traceback (most recent call last):
  File "/Users/macos12/Desktop/Xpra.app/Contents/Resources/lib/python/xpra/net/crypto.py", line 88, in crypto_backend_init
    backend = default_backend()
              ^^^^^^^^^^^^^^^^^
  File "cryptography/hazmat/backends/__init__.pyc", line 11, in default_backend
  File "cryptography/hazmat/backends/openssl/__init__.pyc", line 7, in <module>
  File "cryptography/hazmat/backends/openssl/backend.pyc", line 7, in <module>
ModuleNotFoundError: No module named 'cryptography.hazmat.bindings._rust'
Error: cannot import python-cryptography
```

And I get the same failure when running `python ./xpra/net/net_util.py` from a `jhbuild` shell, so it looks like something is fundamentally broken with the installation of `python-cryptography`'s rust bindings.
There is no `_rust` module in my installation at all!

**Comment by totaam** [comment]
I tried with `pip`, but it moaned:
```
pyopenssl 24.0.0 requires cryptography<43,>=41.0.5, but you have cryptography 43.0.3 which is incompatible.
```
And I eventually landed [here](https://github.com/pyca/pyopenssl/blob/7e1660b84443dae0bfb9e57adcaa2bbd07e5af12/setup.py#L97) which led me to this comment: https://github.com/pyca/pyopenssl/pull/1320#issuecomment-2289146559
Which means that `pyopenssl` and `python-cryptography` have to be updated in sync.
They were not, because `pyopenssl`  was missing from the [dependencies documentation](https://github.com/Xpra-org/xpra/blob/master/docs/Build/Dependencies.md) - now fixed in d9d95f9f8fb3989df5219e973bc939531ec70aeb

`pyOpenSSL` has now been updated in `jhbuild` to version `24.2.1`: https://github.com/Xpra-org/gtk-osx-build/commit/19ea82fc3c275134246de3e869bd30d717250a50 - then another commit because of pypi URL breakage: https://github.com/Xpra-org/gtk-osx-build/commit/a4534f8edcca947ac7ab3d70d01c17b7f7a0a29a

But we're not done yet! Of course not.
It seems that the setuptools build script of `python-cryptography` doesn't actually build the rust bindings!
Oh joy.
```
cd src/rust
cargo build
```
This fails rather dramatically:
```
  = note: Undefined symbols for architecture x86_64:
            "_PyArg_UnpackTuple", referenced from:
                __cffi_f_ASN1_ENUMERATED_set in libcryptography_cffi-f49f72473c4b8b1e.rlib(c4d9326f651c9b94-_openssl.o)
                __cffi_f_ASN1_INTEGER_set in libcryptography_cffi-f49f72473c4b8b1e.rlib(c4d9326f651c9b94-_openssl.o)
                __cffi_f_ASN1_INTEGER_to_BN in libcryptography_cffi-f49f72473c4b8b1e.rlib(c4d9326f651c9b94-_openssl.o)
                __cffi_f_ASN1_STRING_to_UTF8 in libcryptography_cffi-f49f72473c4b8b1e.rlib(c4d9326f651c9b94-_openssl.o)
                __cffi_f_ASN1_TIME_set_string in libcryptography_cffi-f49f72473c4b8b1e.rlib(c4d9326f651c9b94-_openssl.o)
                __cffi_f_ASN1_TIME_to_generalizedtime in libcryptography_cffi-f49f72473c4b8b1e.rlib(c4d9326f651c9b94-_openssl.o)
                __cffi_f_BIO_get_mem_data in libcryptography_cffi-f49f72473c4b8b1e.rlib(c4d9326f651c9b94-_openssl.o)
```
Followed by another few hundred lines of similar missing Python symbols.
And `rustup update` did not help.

That's because of the new python packaging pep517. Lots of emojis in the output, but not very helpful with all the breakage it inflicts. Things were building just fine before this "improvement".

Now, this installs `python-cryptography` correctly on `x86_64`:
```shell
pip install --target  $PREFIX/lib/python3.11/site-packages .
```
(and how are we supposed to call this from `jhbuild`?)

But not on `arm64`:
```
maturin failed
        Caused by: MACOSX_DEPLOYMENT_TARGET is invalid
```
Which is fixed by setting this environment variable before calling pip:
```
export MACOSX_DEPLOYMENT_TARGET=12.0
```

And after all this, the `net_util` tool runs without complaints.
But the bundled application does not:
```py
crypto backend init failure
Traceback (most recent call last):
  File "/Users/totaam/Desktop/Xpra.app/Contents/Resources/lib/python/xpra/net/crypto.py", line 88, in crypto_backend_init
    backend = default_backend()
              ^^^^^^^^^^^^^^^^^
  File "cryptography/hazmat/backends/__init__.pyc", line 11, in default_backend
  File "cryptography/hazmat/backends/openssl/__init__.pyc", line 7, in <module>
  File "cryptography/hazmat/backends/openssl/backend.pyc", line 7, in <module>
  File "cryptography/hazmat/bindings/_rust.pyc", line 13, in <module>
  File "cryptography/hazmat/bindings/_rust.pyc", line 9, in __load
  File "imp.pyc", line 343, in load_dynamic
ImportError: PyO3 modules compiled for CPython 3.8 or older may only be initialized once per interpreter process
Error: cannot import python-cryptography
```
Which looked like a known issue with PyO3: https://github.com/pyca/bcrypt/issues/694 / https://github.com/pyca/cryptography/issues/9016

But it's not!
I got a more helpful error message when trying the import from the bundled Python shell:
```py
$ ./Desktop/Xpra.app/Contents/Helpers/Python          
Python 3.11.9 (main, Sep 14 2024, 17:51:18) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import cryptography
>>> from cryptography.hazmat.backends import default_backend
>>> default_backend()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cryptography/hazmat/backends/__init__.pyc", line 11, in default_backend
  File "cryptography/hazmat/backends/openssl/__init__.pyc", line 7, in <module>
  File "cryptography/hazmat/backends/openssl/backend.pyc", line 7, in <module>
  File "cryptography/hazmat/bindings/_rust.pyc", line 13, in <module>
  File "cryptography/hazmat/bindings/_rust.pyc", line 9, in __load
  File "imp.pyc", line 343, in load_dynamic
ModuleNotFoundError: No module named '_cffi_backend'
```
And so the correct fix is a change to our packaging script instead.
And since I've already spent way too much time on this issue and I couldn't figure out how to make `py2app` bundle `cffi` and `_cffi_backend` properly, so I hacked it instead: be3f3ec6120df4b8872667f0c01c7cdfe4cce362

@psobolewskiPhD please try version `6.3-r36875` or later from the [beta area](https://xpra.org/beta/)

---

BTW, since you're using AES, be aware of some important issues - only some of which will be fully fixed in xpra 6.3: #4375

**Comment by psobolewskiPhD** [comment]
Wow, epic sleuthing.
Will test tomorrow.
To give more context, I'm using xpra in a singularity container in an internal network environement (SLURM cluster) with many users. So it's a pretty secure environment, but want to prevent people from "looking" into colleagues desktops. For now I've switched to using password with a uuidgen string.

**Comment by totaam** [comment]
FYI: finally solved the `arm64` builds properly: #45 and #27

Fresh macos builds will be back soon.
